Allow Nones in Tensor[] for IValues

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143168
* #143167
* #143166

When we parse IValues from Python to C++, we treat Nones as Tensors
(these get parsed into undefined tensors). Before this PR, this logic
only happened for Tensor (and not Tensor[]). For example, we could
create a custom operator foo(Tensor x) and it would accept None as
input.

This PR makes the behavior consistent by extending it to Tensor[].

I need this behavior later on when parsing lists of grad_output which
look like they are List[Tensor] in C++ but become List[Optional[Tensor]]
in Python (because undefined tensor in C++ maps to Python None).

Test Plan:
- new test

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

zou3519 committed:
Allow Nones in Tensor[] for IValues

When we parse IValues from Python to C++, we treat Nones as Tensors
(these get parsed into undefined tensors). Before this PR, this logic
only happened for Tensor (and not Tensor[]). For example, we could
create a custom operator foo(Tensor x) and it would accept None as
input.

This PR makes the behavior consistent by extending it to Tensor[].

I need this behavior later on when parsing lists of grad_output which
look like they are List[Tensor] in C++ but become List[Optional[Tensor]]
in Python (because undefined tensor in C++ maps to Python None).

Test Plan:
- new test

[ghstack-poisoned]

ezyang reviewed (approved):
How come the optional wraps the vector as opposed to other way around?

zou3519 commented:
> How come the optional wraps the vector as opposed to other way around?

whoops I intended for it to be `vector<optional<Tensor>>`, thanks for the catch.

zou3519 committed:
Update on "Allow Nones in Tensor[] for IValues"


When we parse IValues from Python to C++, we treat Nones as Tensors
(these get parsed into undefined tensors). Before this PR, this logic
only happened for Tensor (and not Tensor[]). For example, we could
create a custom operator foo(Tensor x) and it would accept None as
input.

This PR makes the behavior consistent by extending it to Tensor[].

I need this behavior later on when parsing lists of grad_output which
look like they are List[Tensor] in C++ but become List[Optional[Tensor]]
in Python (because undefined tensor in C++ maps to Python None).

Test Plan:
- new test

cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]
